kwave/utils/mapgen.py

```diff
@@ -2863,6 +2863,117 @@ def calculate_lateral_pressure() -> NDArray[Shape["N"], Float]:
     return p_axial, p_lateral, p_axial_complex
 
 
+@beartype
+def focused_annulus_oneil(radius: float,
+                          diameter: kt.NP_ARRAY_FLOAT_2D,
+                          amplitude: kt.NP_ARRAY_FLOAT_1D,
+                          phase: kt.NP_ARRAY_FLOAT_1D,
+                          frequency: kt.NUMERIC,
+                          sound_speed: kt.NUMERIC,
+                          density: kt.NUMERIC,
+                          axial_positions: Union[kt.NP_ARRAY_FLOAT_1D, float, list]) -> Union[kt.NP_ARRAY_FLOAT_1D, float]:
+    """Compute axial pressure for focused annulus transducer using O'Neil's solution
+
+    focused_annulus_oneil calculates the axial pressure for a focused
+    annular transducer using O'Neil's solution (O'Neil, H. Theory of
+    focusing radiators. J. Acoust. Soc. Am., 21(5), 516-526, 1949). The
+    annuluar elements are uniformly driven by a continuous wave sinusoid
+    at a given frequency and normal surface velocity.
+
+    The solution is evaluated at the positions (along the beam axis) given
+    by axial_position. Where 0 corresponds to the transducer surface.
+
+    Note, O'Neil's formulae are derived under the assumptions of the
+    Rayleigh integral, which are valid when the transducer diameter is
+    large compared to both the transducer height and the acoustic
+    wavelength.
+
+    Example:
+        # define transducer parameters
+        radius = 140e-3  # [m]
+        diameter = 120e-3  # [m]
+        velocity = 100e-3  # [m / s]
+        frequency = 1e6  # [Hz]
+        sound_speed = 1500  # [m / s]
+        density = 1000  # [kg / m^3]
+
+        # define position vectors
+        axial_position = np.arange(0, 250e-3 + 1e-4, 1e-4)  # [m]
+        p_axial = focused_annulus_oneil(radius, diameter, amplitude, phase, frequency, sound_speed, density, axial_position)
+
+    Args:
+        radius: transducer radius of curvature [m]
+        diameter: 2 x num_elements array containing pairs of inner and outer aperture diameter
+                  (diameter of opening) [m].
+        amplitude: array containing the normal surface velocities for each element [m/s]
+        phase: array containing the phase for each element [rad]
+        frequency: driving frequency [Hz]
+        sound_speed: speed of sound in the propagating medium [m/s]
+        density: density in the propagating medium [kg/m^3]
+        axial_positions: vector of positions along the beam axis where the
+                         pressure amplitude is calculated [m]
+
+    Returns:
+        p_axial: pressure amplitude at the positions specified by axial_position [Pa]
+
+    References:
+        O'Neil, H. (1949). Theory of focusing radiators. J. Acoust. Soc. Am., 21(5), 516-526.
+
+    See also focused_bowl_oneil.
+    """
+
+    assert (diameter.ndim == 2), "wrong shape for diameter"
+    assert (2 in np.shape(diameter)), "wrong shape for diameter"
```

Comment:
This assertion is covered by the beartyping and will therefore never be reached.